python/convert.c

```diff
@@ -0,0 +1,344 @@
+  // or __int__()) that users expect.
+  *val = PyLong_AsLongLong(obj);
+  if (!PyErr_Occurred()) return true;
+  if (PyErr_ExceptionMatches(PyExc_OverflowError)) {
+    // Rewrite OverflowError -> ValueError.
+    // But don't rewrite other errors such as TypeError.
+    PyErr_Clear();
+    PyErr_Format(PyExc_ValueError, "Value out of range: %S", obj);
+  }
+  return false;
+}
+
+static bool PyUpb_GetUint64(PyObject *obj, uint64_t *val) {
+  // For uint64 Python does not offer any functions as convenient as
+  // PyLong_AsLongLong().  If the object is not already a "long" we must
+  // manually perform the automatic conversion (using __index__() or __int__())
+  // that users expect.
+  if (PyLong_Check(obj)) {
+    *val = PyLong_AsUnsignedLongLong(obj);
+  } else {
+    PyObject* casted = PyNumber_Long(obj);
+    if (!casted) return false;
+    *val = PyLong_AsUnsignedLongLong(casted);
+    Py_DECREF(casted);
+  }
+  if (!PyErr_Occurred()) return true;
+  PyErr_Clear();
+  PyErr_Format(PyExc_ValueError, "Value out of range: %S", obj);
+  return false;
+}
+
+// If `arena` is specified, copies the string data into the given arena.
+// Otherwise aliases the given data.
+static upb_msgval PyUpb_MaybeCopyString(const char *ptr, size_t size,
+                                        upb_arena *arena) {
+  upb_msgval ret;
+  ret.str_val.size = size;
+  if (arena) {
+    char *buf = upb_arena_malloc(arena, size);
+    memcpy(buf, ptr, size);
+    ret.str_val.data = buf;
+  } else {
+    ret.str_val.data = ptr;
+  }
+  return ret;
+}
+
+static bool PyUpb_PyToUpbEnum(PyObject *obj, const upb_enumdef *e,
+                              upb_msgval *val) {
+  if (PyUnicode_Check(obj)) {
+    Py_ssize_t size;
+    const char *name = PyUnicode_AsUTF8AndSize(obj, &size);
+    const upb_enumvaldef *ev = upb_enumdef_lookupname(e, name, size);
+    if (!ev) {
+      PyErr_Format(PyExc_ValueError, "unknown enum label \"%s\"", name);
+      return false;
+    }
+    val->int32_val = upb_enumvaldef_number(ev);
+    return true;
+  } else {
+    if (!PyUpb_GetInt64(obj, &val->int64_val) ||
+        val->int64_val < INT32_MIN || val->int64_val > INT32_MAX) {
+      return false;
+    }
+    val->int32_val = val->int64_val;
+    if (upb_filedef_syntax(upb_enumdef_file(e)) == UPB_SYNTAX_PROTO2 &&
+        !upb_enumdef_checknum(e, val->int32_val)) {
+      PyErr_Format(PyExc_ValueError, "invalid enumerator %d",
+                    (int)val->int32_val);
+      return false;
+    }
+    return true;
+  }
+}
+
+bool PyUpb_PyToUpb(PyObject *obj, const upb_fielddef *f, upb_msgval *val,
+                   upb_arena *arena) {
+  switch (upb_fielddef_type(f)) {
+    case UPB_TYPE_ENUM:
+      return PyUpb_PyToUpbEnum(obj, upb_fielddef_enumsubdef(f), val);
+    case UPB_TYPE_INT32:
+      if (!PyUpb_GetInt64(obj, &val->int64_val) || val->int64_val < INT32_MIN ||
+          val->int64_val > INT32_MAX) {
+        return false;
+      }
+      val->int32_val = val->int64_val;
```

Comment:
this replicates the error checking code of the enum above.  Maybe share something?

Reply:
Done.